### ISSUE TITLE
What does optax.mask do?

### ISSUE THREAD

**Issue by PhilipVinc** [deepmind/optax#159]
I don't understand what `optax.mask` does.

I would expect that the masked optimizer
```python
optax.mask(optax.sgd(0.1), {"Dense": True, "bias": False})
```
would only apply the optimisation to sub-leafs of `Dense` and return `0` for `bias`.
However it seems to me that the masked updates are correct for sub-leafs of `Dense` (so where the mask is True, but they are the identity where the mask is False.

Is this intended behaviour? it seems rather strange to me.
I was trying to update only a subsets of the weights of my model but this was not working

MWE: 
```python
import jax.numpy as jnp
import jax
import optax

pars = {"Dense": {"kernel": jnp.zeros((2,3)), "bias": jnp.zeros((3))}, "bias":jnp.zeros(2)}
grad = jax.tree_map(jnp.ones_like, pars)

op = optax.masked(optax.sgd(0.1), {"Dense": True, "bias": False})

op_state = op.init(pars)

masked_updates, new_op_state = op.update(grad, op_state,  pars)
>>> masked_updates
{'Dense': {'bias': DeviceArray([-0.1, -0.1, -0.1], dtype=float32), 'kernel': DeviceArray([[-0.1, -0.1, -0.1],
             [-0.1, -0.1, -0.1]], dtype=float32)}, 'bias': DeviceArray([1., 1.], dtype=float32)}

```

**Comment by pharringtonp19** [comment]
@PhilipVinc  Any success on understanding why this behavior occurs?

**Comment by n2cholas** [comment]
This is intended behaviour. `optax.masked` controls which gradients the inner transformation will affect. In your example, you pass in gradients that are all ones. `optax.masked(optax.sgd(0.1), {"Dense": True, "bias": False})` says apply the `sgd` transformation to the `Dense` gradients, but not the `bias` gradients. One motivating use case is applying weight decay to only some of your gradients. 

There is currently no transformation that will zero out gradients for you, but this can be done pretty simply with `multi_transform`:

```py
import jax.numpy as jnp
import jax
import optax

pars = {"Dense": {"kernel": jnp.zeros((2,3)), "bias": jnp.zeros((3))}, "bias": jnp.zeros(2)}
grad = jax.tree_map(jnp.ones_like, pars)

def zero_grads():
    def init_fn(_): 
        return ()
    def update_fn(updates, state, params=None):
        return jax.tree_map(jnp.zeros_like, updates), ()
    return optax.GradientTransformation(init_fn, update_fn)

op = optax.multi_transform({"sgd": optax.sgd(0.1), "zero": zero_grads()},
                           {"Dense": "sgd", "bias": "zero"})
op_state = op.init(pars)

masked_updates, new_op_state = op.update(grad, op_state,  pars)
```

Does that help? Do you have suggestions for how to improve the docstring for `masked` to make this more clear?

**Comment by PhilipVinc** [comment]
@pharringtonp19 sorry i was on holiday.
@n2cholas I eventually had figured this out. I admit, this behaviour makes complete sense within the framework of `gradient transformation` of optax, however it took me a while to figure it out.

the current docstrings states `Mask updates so only a subset of them are computed.` .
In optax view, that means that only a subset of the gradient are transformed, and the rest arent.
But if one isn't familiar with optax, i would immediately think that only some gradients are computed, and the rest are zeroed out or whatever.

I think the docstring could be modified to state something like:

```python
  """Mask updates so only a subset of them are transformed, while the rest are 
passed-through unchanged.
"""
```

**Comment by PhilipVinc** [comment]
I will tomorrow

**Comment by pharringtonp19** [comment]
Out of curiosity here, when we `jit' the training loop, are the gradients that we zero out computed? I am wondering whether XLA realizes that we don't need those gradients and therefore doesn't compute them in the first place.

Is there a simple way to check this?

**Comment by PhilipVinc** [comment]
I think it does, because Dead Code Elimination will see that they are not used anywhere. (`zero_grad` calls `zeros_like`)
